### PR TITLE
[WIP] port gstdec to gstreamer 1.x

### PR DESCRIPTION
**Pull request by jrobeson**
This PR is still a work in progress.

Still need to figure out how to get the buffer data.
I've been testing this with decode.py and I end up with:

TypeError: can't convert return value to desired type

Note: I don't actually know much about how Gstreamer works internally,
so definitely take a good look.

### CODE REVIEW

**Comment by jrobeson** [comment]
It's kinda hard to debug this with the threading. I tried using pdb to step through it, but I don't really know what I'm seeing yet. Here's me stepping through (After the file read):

http://fpaste.org/226431/28044051/

File: audioread/gstdec.py [hunk 1]
```diff
@@ -47,9 +47,11 @@
 """
 from __future__ import with_statement
 
-import gst
+import gi
+gi.require_version('Gst', '1.0')
```
Comment (sampsyo) [inline audioread/gstdec.py]: Maybe this test (and `Gst.init(None)`, below) should be done on-demand, so they can't fail after the `_gst_available` check succeeds?

**Comment by sampsyo** [comment]
Weirdly, I'm not getting a TypeError—I'm getting the code's own MetadataMissingError suggesting that the duration couldn't be determined.

```
Traceback (most recent call last):
  File "decode.py", line 52, in <module>
    decode(sys.argv[1])
  File "decode.py", line 31, in decode
    with audioread.gstdec.GstAudioFile(filename) as f:
  File "/home/asampson/audioread/audioread/gstdec.py", line 248, in __init__
    raise self.read_exc
audioread.gstdec.MetadataMissingError: duration not available
```

What does your traceback look like?

**Comment by jrobeson** [comment]
I wish i was getting a traceback :).  I'm just seeing "TypeError: can't convert return value to desired type" over and over again. From my understanding, that means the problem is happening in a callback.

I'm just using the code from the gstdec smoke test copied to info.py
```python
from __future__ import print_function
import audioread
import sys
import os
import contextlib

if __name__ == '__main__':
    for path in sys.argv[1:]:
        path = os.path.abspath(os.path.expanduser(path))
        with audioread.audio_open(path) as f:
            print('Backend:', str(type(f).__module__).split('.')[1],
                  file=sys.stderr)
            print(f.channels)
            print(f.samplerate)
            print(f.duration)

            for s in f:
                print(len(s), ord(s[0]))
```

and ending up with:
```bash
$ python info.py foo.mp3
Backend: gstdec
2
44100
280.764081632
can't convert return value to desired type
TypeError: can't convert return value to desired type4608 197

TypeError: 4608 76
can't convert return value to desired type
TypeError: 4608 231
can't convert return value to desired type
TypeError: can't convert return value to desired type4608 81

TypeError: can't convert return value to desired type4608 164

```

I do see the duration  though.

**Comment by jrobeson** [comment]
What file did you use? What version of python are you testing against?  I'm using python 2.7.9 and gstreamer 1.4.5 on Fedora 22.

**Comment by sampsyo** [comment]
That's super strange; I wish we knew where the error was being raised!

Good questions.
* Python 2.7.9
* Arch Linux
* gstreamer 1.4.5
* I used this tiny MP3: https://github.com/sampsyo/beets/blob/master/test/rsrc/full.mp3

**Comment by jrobeson** [comment]
Here's my results with the same script as above:
```bash
[johnny@falling ~/src/audioread (port-gstdec-to-gstreamer1 *)]$ python info.py full.mp3 
Backend: ffdec
1
44100
1.0
```

**Comment by jrobeson** [comment]
I can replicate your duration not available with that mp3 file. probably because it's so short. Thanks for that, I'll play with it!

**Comment by sampsyo** [comment]
Interesting! Seems like a race condition, then?

**Comment by jrobeson** [comment]
I did verify that it can in fact get to the point where we can pull out the channels and sample rate, so it can definitely read the file. It's just that the Gst query fails.

I can't say I'm too familiar with debugging threaded applications.  It'd be a lot simpler if we could just setup a pipeline to dump to wave directly instead of relying on the wave module.

**Comment by sampsyo** [comment]
Indeed, mostly because GStreamer *needs* to run in its own thread—and is mainly geared toward real-time decoding. :confused: 

So are you concluding that the channels, etc. *eventually* appear but aren't necessarily ready immediately? If so, maybe we just need to figure out the right synchronization. I can look into that.

**Comment by jrobeson** [comment]
I think so. All i did was was add a print() for 

```python
        self.channels = info.get_int('channels')[1]
        self.samplerate = info.get_int('rate')[1]
```

and that shows 1 and 44100 respectively

**Comment by jrobeson** [comment]
Someone in #gstreamer (on freenode) suggested that that gstreamer might have a bug somewhere related to ultra short durations. They tested with an mp3 file built with this pipeline
```
 audiotestsrc num-buffers=10 ! lamemp3enc ! filesink location=test.mp3
```
and got a duration of 0 seconds

If it turns out to be a bug, then I'll point out a link to it.

For now though, can you try this with a slighty longer mp3? Maybe you can solve the race condition I'm seeing.

**Comment by sampsyo** [comment]
You're absolutely right; this worked with a longer MP3.

I found and fixed the problem with the return value type. I pulled this into a branch and opened a new PR, #23. I'm also going to add you as a collaborator, so please feel free to work in the branch on this repository so we can both commit now.

I'm not sure whether the short-file issue was present in the old gstreamer 0.10 version—or if there's anything to be done to work around it. Other than that, though, maybe this is ready? It would be worth testing a little more.

**Pull request closed without merging**